xds/src/main/java/io/grpc/xds/WeightedTargetLoadBalancer.java

```diff
@@ -0,0 +1,216 @@
+      } else if (!weightedChildLbConfig.policyName.equals(targets.get(targetName).policyName)) {
+        // Policy name change for the same target is not a typical usecase, but we should support
+        // it. Switch child policy.
+        childBalancers.get(targetName)
+            .switchTo(lbRegistry.getProvider(weightedChildLbConfig.policyName));
+      }
+    }
+
+    // Update new config map.
+    targets = newTargets;
+
+    // Call handleResolvedAddresses() for each child balancer.
+    for (String targetName : targets.keySet()) {
+      childBalancers.get(targetName).handleResolvedAddresses(
+          resolvedAddresses.toBuilder()
+              .setLoadBalancingPolicyConfig(targets.get(targetName).config)
+              .build());
+    }
+
+    // Cleanup removed targets.
+    // TODO(zdapeng): cache removed target for 15 minutes.
+    for (String targetName : childBalancers.keySet()) {
+      if (!targets.containsKey(targetName)) {
+        childBalancers.get(targetName).shutdown();
+      }
+    }
+    childBalancers.keySet().retainAll(targets.keySet());
+    childHelpers.keySet().retainAll(targets.keySet());
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    logger.log(XdsLogLevel.WARNING, "Received name resolution error: {0}", error);
+    if (childBalancers.isEmpty()) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+    for (LoadBalancer childBalancer : childBalancers.values()) {
+      childBalancer.handleNameResolutionError(error);
+    }
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (LoadBalancer childBalancer : childBalancers.values()) {
+      childBalancer.shutdown();
+    }
+  }
+
+  private void updateBalancingState() {
+    List<WeightedChildPicker> childPickers = new ArrayList<>();
+
+    ConnectivityState overallState = null;
+    for (String name : targets.keySet()) {
+      ChildHelper childHelper = childHelpers.get(name);
+      ConnectivityState childState = childHelper.currentState;
+      overallState = aggregateState(overallState, childState);
+      if (READY == childState) {
+        int weight = targets.get(name).weight;
+        childPickers.add(new WeightedChildPicker(weight, childHelper.currentPicker));
+      }
+    }
+
+    SubchannelPicker picker;
+    if (childPickers.isEmpty()) {
+      if (overallState == TRANSIENT_FAILURE) {
+        picker = new ErrorPicker(Status.UNAVAILABLE); // TODO: more details in status
+      } else {
+        picker = XdsSubchannelPickers.BUFFER_PICKER;
+      }
+    } else {
+      picker = weightedPickerFactory.picker(childPickers);
+    }
+
+    if (overallState != null) {
+      helper.updateBalancingState(overallState, picker);
+    }
+  }
+
+  @Nullable
+  private ConnectivityState aggregateState(
```

Comment:
We may want to make a utility method for this, as it is used in a couple of places (for any load balancers that contain a bunch of child balancers).

Reply:
This is related to #6650. Might do it after the design in general is clear.